backend/src/oauth/AuthorizationCodeModelImpl.ts

```diff
@@ -0,0 +1,394 @@
+    try {
+      const {
+        aud,
+        exp,
+        iat,
+        sub,
+        scope,
+        type
+      } = jwt.verify(refreshToken, process.env.JWT_SECRET_KEY) as JwtPayload;
+
+      const clientId = aud.toString();
+      const userId = parseInt(sub, 10);
+
+      // Verify token validity (returns Error object on error and grants on success)
+      const result = await AuthorizationCodeModelImpl.verifyTokenValidity(
+        this.repo,
+        type,
+        TokenType.Refresh,
+        exp,
+        iat,
+        clientId,
+        userId
+      );
+      if (result instanceof Error) {
+        throw result;
+      }
+
+      const decoded: RefreshToken = {
+        refreshToken,
+        refreshTokenExpiresAt: new Date(exp * 1000),
+        scope,
+        client: {
+          id: clientId,
+          grants: result ? result : []
+        },
+        user: { id: userId }
+      };
+      if (callback) {
+          callback(null, decoded);
+      }
+      return decoded;
+
+    } catch (error) {
+      // Error handling: token may be expired or have an invalid signature.
+      if (callback) {
+        callback(error, null);
+      }
+      return null;
+    }
+  }
+
+  async revokeToken(token: RefreshToken | Token, callback?: Callback<boolean>): Promise<boolean> {
+    // revocation of the individual tokens is not supported
+    if (callback) {
+      callback(null, true);
+    }
+    return true;
+  }
+
+  async generateAuthorizationCode(
+    client: Client,
+    user: User,
+    scope: string | string[],
+    callback?: (err: Error | null, authorizationCode?: string) => void
+  ): Promise<string> {
+    const authCode = randomBytes(32).toString('hex');
+
+    const scopeStr = Array.isArray(scope) ? scope.join(' ') : scope;
+    await this.repo.saveOrUpdateConsent(client.id, user.id, scopeStr);
+
+    if (callback && typeof callback === 'function') {
+      callback(null, authCode);
+    }
+    return authCode;
+  }
+
+  static generateJwtToken(
+      sub: string,
+      exp: number,
+      iat: number,
+      aud: string,
+      scope: string | string[],
+      type: TokenType,
+      additionalFields: Record<string, unknown> = {}
+  ): string {
+    const iss = 'https://orbitar.space';
+    scope = Array.isArray(scope) ? scope.join(' ') : scope;
+    const payload = {
+      aud,
+      iss,
+      exp,
+      iat,
+      sub,
+      scope,
+      type,
+      ...additionalFields
+    };
+
+    return jwt.sign(payload, process.env.JWT_SECRET_KEY);
+  }
+
+  static generateClientId = (): string => {
+    return crypto.randomUUID();
+  };
+
+  static generateClientSecret = (): string => {
+    return crypto.randomBytes(32).toString('hex');
+  };
+
+  static hashString(value: string): string {
+    return crypto.createHash('sha256').update(value).digest('hex');
+  }
+
+  static async verifyTokenValidity(
+    repo: OAuth2Repository,
```

Comment:
why not make it non-static to avoid passing repo?

Reply:
easier to write tests for it, but I haven't yet checked if we have db layer in tests, if not, I will make it non-static